### PR TITLE
Export individual nodes

### PR DESCRIPTION
**Pull request by ebrahimebrahim**
This is my second go at addressing #4989. Related links:
- [first try](https://github.com/Slicer/Slicer/pull/5805)
- [my comment](https://github.com/Slicer/Slicer/issues/4989#issuecomment-904779802) that followed discussion about the first PR
- [related SlicerMorph issue](https://github.com/SlicerMorph/SlicerMorph/issues/72)

This time the export module is a loadable module written in C++, and it gets its own `qSlicerFileDialog` that is opened via `qSlicerApplication::application()->ioManager()->openDialog`.

@lassoan @pieper @cpinter @smrolfe  @sjh26  

# Demo

Right click on an item in the subject hierarchy view, and there will be an "Export" action available if it is a storable node. This brings up a simple export dialog.

sc1

sc2

# Commits

I will squash all "WIP" commits into one "ENH" commit if this becomes ready to be merged. The non-WIP commits were minor unrelated things that I'll keep as separate commits.

# Choices made

*Reporting various errors while saving:* Where the save dialog has error/warning icons with tooltips containing error messages, I went with `QMessageBox`s to display error messages. I think it makes sense here, since it's just one node that the errors can be for, and since users would probably expect a message box in the failure case for a single task. The error messages are sometimes not pretty, because they are often just a dump of what is in `storageNode->GetUserMessages()`. But they get the job done.

sc3

# Remaining problems, missing features

There are some things I am leaving to be addressed while this PR is looked at, or afterwards in a different PR.

1. ~~Last used export settings (file format, options, and directory) for a given node type should be remembered during the session. This shouldn't be difficult to do.~~ **Update: This is done.**
2. Selecting multiple nodes of the same type, or a folder with child nodes all the same time, should provide the export action. This is a big feature that would be in a separate PR. **Update: The non-GUI part of this is actually done, so implementing it later will not be an API change.**
3. ~~From [here](https://github.com/SlicerMorph/SlicerMorph/issues/72) I'm not sure how to allow the user to specify a coordinate system. My first thought is that a checkbox+tooltip should be added to `qSlicerNodeWriterOptionsWidget` and made visible for certain nodes, but I would definitely need help with making it do the right thing. We can look into this later.~~
    **Update: This is done. There's now a checkbox for transform hardening.**
4. Is there some way in which exporting needs to be linked up to the python side of things in order to be extensible by scripting? I don't know much about this!
5. Some formats are shown that don't make sense, e.g. ".bmp" as an export format for 3D volumes. The problem here comes from the list of extensions provided by [this](https://github.com/Slicer/Slicer/blob/bcaab0a249d444d55aa6fc808bd43837cd8f5db7/Base/QTCore/qSlicerCoreIOManager.h#L72), and it's a problem that is shared with the scene save dialog. Should be looked at in a different PR, or opened as an issue on its own.
6. I'm not sure that the compress option should really be shown as often as it is. For a format like `.nrrd`, the compress option clearly does something. However with `.nii` and `.nii.gz` I would think that whether or not there is compression is built into the choice of file type. If that's the case then a compress option shouldn't be displayed for these, but I don't really know how it works.
7. ~~You might see the visibility of the "compress" checkbox for segmentation exports randomly fluctuate from time to time. It happened in the main save dialog too. This is due to an uninitialized variable bug [here](https://github.com/Slicer/Slicer/blob/bcaab0a249d444d55aa6fc808bd43837cd8f5db7/Modules/Loadable/Segmentations/qSlicerSegmentationsNodeWriter.cxx#L51). Should be an easy fix, and I can include it in this PR if that seems fine.~~ **Update: This is done in 565ba958bdba265de78314c888aa865de3d00025.**
8. ~~*Typing a filename and hitting enter sometimes has unintuitive behavior.* When I type a filename into an export dialog and hit *enter*, then if the dialog closes what I expect is that the file was saved with the exact name that I typed in. Currently that's not always the case, and I think it's not acceptable for it to remain this way. For example, if I type "MRHead.NRRD" and hit enter, it will save as "MRHead.nrrd". If I type "MRHead.nrr" while the "(.nrrd)" option is selected in the format dropdown menu, then hitting enter will save the file as "MRHead.nrr.nrrd". This is just imitating the filename editing behavior in the main save dialog, but it was okay in the main save dialog because hitting *enter* while editing a filename just finalized the edit in the `QTableWidget` cell. I'm facing a conflict here between following the way the main save dialog handles things (for the sake of consistent behavior across the application) and avoiding unintuitive/intrusive behaviors. To clarify: if the user clicks out of the filename editing box, then they will see the filename being corrected. The problem here is that they can hit *enter* without ever clicking out of the box. Ideas:~~
	- ~~Don't change the user-given filename, and let them get away with anything (besides characters that are not legal in filenames). If they want to export to a nifti file and call it "MRHead.bmp," then they can do that.~~
	- ~~Don't change the user-given filename, but if they try to export to a nifti file named "MRHead.bmp", then they get a prompt offering to set a better name, which they can answer "yes/no/cancel".~~
	**Update: This is handled now, via the latter idea.** 
9. Somewhat related to above, and this is a problem with the main save dialog too: It's a little weird that we display differently-cased extensions as separate items in the format list (e.g. ".png" and ".PNG" are separate entries), but we only accept an extension when the user types it in lowercase (e.g. trying to use "MRHead.PNG" as a filename can result in something like "MRHead.PNG.png"). It should really be the other way around: only display lowercase in the format dropdown menu, but accept any case the user chooses to type in. Maybe this should be a separate PR/issue.

### CODE REVIEW

**Comment by pieper** [comment]
> 1. Last used export settings (file format, options, and directory) for a given node type should be remembered during the session. This shouldn't be difficult to do.

Yes, I think this should be done.  It was on @lassoan 's wishlist for moving this feature to the core.

> 2. Selecting multiple nodes of the same type, or a folder with child nodes all the same time, should provide the export action. This is a big feature that would be in a separate PR.

Yes, make this a different PR since this is already long.

> 3. From [here](https://github.com/SlicerMorph/SlicerMorph/issues/72) I'm not sure how to do this one:
    > need a way to specify if user wants the node to be exported in world coordinate system (parent transform hardened) or in the node's coordinate system (without applying the parent transform)

>    My first thought is that a checkbox+tooltip should be added to `qSlicerNodeWriterOptionsWidget` and made visible for certain nodes, but I would definitely need help with making it do the right thing. We can look into this later.

This is also important for many use cases and already works in the ExportAs python version.  Fortunately I think that with your current approach ExportAs will still work, so it can provide this feature in SlicerMorph (but it will be awkward to have both).  Better to implement as part of this feature I if possible so ExportAs can be retired completely.

The code should be identical to the python version; did you try to implement it?  I think yes a checkbox on the dialog for hardening makes sense and you can know whether to display it by checking if the selected node is a `vtkMRMLTransformableNode`.

> 4. Is there some way in which exporting needs to be linked up to the python side of things in order to be extensible by > scripting? I don't know much about this!

> 5. Some formats are shown that don't make sense, e.g. ".bmp" as an export format for 3D volumes. The problem here comes from the list of extensions provided by [this](https://github.com/Slicer/Slicer/blob/bcaab0a249d444d55aa6fc808bd43837cd8f5db7/Base/QTCore/qSlicerCoreIOManager.h#L72), and it's a problem that is shared with the scene save dialog. Should be looked at in a different PR, or opened as an issue on its own.

Yes, fixing this at a lower level is better.

> 6. I'm not sure that the compress option should really be shown as often as it is. For a format like `.nrrd`, the compress option clearly does something. However with `.nii` and `.nii.gz` I would think that whether or not there is compression is built into the choice of file type. If that's the case then a compress option shouldn't be displayed for these, but I don't really know how it works.

Having the compression option is very important for many cases.  For large complex data there can be a 10x time penalty to get a 30% space savings.

If there are formats where it doesn't make sense we should expose that attribute at a lower level and only show the checkbox when it makes sense.

> 7. You might see the visibility of the "compress" checkbox for segmentation exports randomly fluctuate from time to time. It happened in the main save dialog too. This is due to an uninitialized variable bug [here](https://github.com/Slicer/Slicer/blob/bcaab0a249d444d55aa6fc808bd43837cd8f5db7/Modules/Loadable/Segmentations/qSlicerSegmentationsNodeWriter.cxx#L51). Should be an easy fix, and I can include it in this PR if that seems fine.

Definitely let's fix that one.

> 8. *Typing a filename and hitting enter sometimes has unintuitive behavior.* When I type a filename into an export dialog and hit *enter*, then if the dialog closes what I expect is that the file was saved with the exact name that I typed in. Currently that's not always the case, and I think it's not acceptable for it to remain this way. For example, if I type "MRHead.NRRD" and hit enter, it will save as "MRHead.nrrd". If I type "MRHead.nrr" while the "(.nrrd)" option is selected in the format dropdown menu, then hitting enter will save the file as "MRHead.nrr.nrrd". This is just imitating the filename editing behavior in the main save dialog, but it was okay in the main save dialog because hitting *enter* while editing a filename just finalized the edit in the `QTableWidget` cell. I'm facing a conflict here between following the way the main save dialog handles things (for the sake of consistent behavior across the application) and avoiding unintuitive/intrusive behaviors. To clarify: if the user clicks out of the filename editing box, then they will see the filename being corrected. The problem here is that they can hit *enter* without ever clicking out of the box. Ideas:
>	- Don't change the user-given filename, and let them get away with anything (besides characters that are not legal in filenames). If they want to export to a nifti file and call it "MRHead.bmp," then they can do that.
>	- Don't change the user-given filename, but if they try to export to a nifti file named "MRHead.bmp", then they get a prompt offering to set a better name, which they can answer "yes/no/cancel".

The second option makes sense - only raise the dialog in cases where you think the default behavior is probably a mistake.

> 9. Somewhat related to above, and this is a problem with the main save dialog too: It's a little weird that we display differently-cased extensions as separate items in the format list (e.g. ".png" and ".PNG" are separate entries), but we only accept an extension when the user types it in lowercase (e.g. trying to use "MRHead.PNG" as a filename can result in something like "MRHead.PNG.png"). It should really be the other way around: only display lowercase in the format dropdown menu, but accept any case the user chooses to type in. Maybe this should be a separate PR/issue.

I agree this is odd and should be fixed at a lower level to it fixes both dialogs.  Can be a different PR.

**Comment by ebrahimebrahim** [comment]
Thanks @pieper for all the comments!! 

Thanks for letting me know where to look for the transform-hardening option.

I'll be chipping away at some of these items for now.

**Comment by ebrahimebrahim** [comment]
We discussed this PR in the 9/28 slicer hangout.
(Summary of comments](https://discourse.slicer.org/t/2021-09-28-weekly-meeting/19885/8?u=ebrahim)

**Comment by ebrahimebrahim** [comment]
@lassoan @pieper @jcfr This is now ready to be reviewed.

From the items that I listed in my original post above, I completed the ones that need to be in this PR. I plan to open issues for the others, once this is merged.

File: Base/QTGUI/Resources/UI/qSlicerExportNodeDialog.ui
Comment (lassoan) [inline Base/QTGUI/Resources/UI/qSlicerExportNodeDialog.ui:101]: "Harden" could mislead the users to think that the transform is hardened on the node (and thereby modifying the node in the scene.
```suggestion
        <string>Apply transforms</string>
```

**Review by lassoan — CHANGES_REQUESTED** [review]
Thanks for your contribution. This works great!

A few findings:
- Error is not displayed. For example, set "Directory" to a non-existing drive or read-only location (x:\tmp or C:\Windows) and click Save. The saving obviously fails, but no error is reported. On models, the error is just silent. For volumes, Slicer crashes.
- Only show "Harden transform" option if the node has a parent transform.
- see more comments inline

File: Base/QTGUI/qSlicerExportNodeDialog.cxx
Comment (ebrahimebrahim) [reply Base/QTGUI/qSlicerExportNodeDialog.cxx:45]: Good point; what would be a good way to organize this?

Initially I was `#include`ing the `qSlicerSaveDataDialog` header to access these functions, but I preferred to have zero coupling between `qSlicerSaveDataDialog` and `qSlicerExportNodeDialog`.

If I move these utility functions to a third place that can be accessed by both, what form should that take? A header only, or a header and source? Would it live in `Base/QTGUI`?

File: Base/QTGUI/qSlicerExportNodeDialog.cxx
Comment (ebrahimebrahim) [reply Base/QTGUI/qSlicerExportNodeDialog.cxx:522]: Yes that's what I mean. I mean to say that I'm relying on the memory management built into QWidget hierarchies. The widget returned by `qSlicerCoreIOManager::fileWriterOptions` in

```cpp
qSlicerIOOptions* options =
    coreIOManager->fileWriterOptions(this->MRMLNode, this->formatText());
```

has no owner managing its memory.

Maybe it would be clearer if I made the comment say "we will pass ownership of optionsWidget to the qSlicerExportNodeDialog widget" ?

File: Base/QTCore/qSlicerCoreIOManager.cxx
Comment (ebrahimebrahim) [reply Base/QTCore/qSlicerCoreIOManager.cxx:756]: Can you describe an example scenario where transform hardening reuses existing data objects in the node? If possible I'd like to test out a scenario where this happens.

Comment (lassoan) [reply Base/QTCore/qSlicerCoreIOManager.cxx:756]: A shallow copy does not create an independent copy of an object but some data can be shared. It is up to the node to decide if all data is shared or nothing. If you change anything in the destination object in place (e.g., model point coordinates) and that data happens to be shared with the source object then the source object gets modified, too.

File: Base/QTGUI/qSlicerExportNodeDialog.cxx
Comment (lassoan) [reply Base/QTGUI/qSlicerExportNodeDialog.cxx:522]: > Maybe it would be clearer if I made the comment say "we will pass ownership of optionsWidget to the qSlicerExportNodeDialog widget" ?

Yes

File: Base/QTGUI/qSlicerExportNodeDialog.cxx
Comment (lassoan) [reply Base/QTGUI/qSlicerExportNodeDialog.cxx:45]: For minimizing the amount of change, the simplest is to move it as static methods to a class that they both use. Maybe qSlicerCoreIOManager or something similar.

File: Base/QTCore/qSlicerCoreIOManager.cxx
Comment (ebrahimebrahim) [reply Base/QTCore/qSlicerCoreIOManager.cxx:756]: Right, I was thinking that this is desirable behavior. But I see now-- you're saying that the transform hardening may modify the underlying data in the temporary node. It didn't in the cases I tested, because I was only messing around with linear transforms. I guess linear transforms on a volume only affect the `vtkImageData` axis orientations and not the underlying data array. I should test this out with transforms that modify the underlying data.

Okay, so make sure to deep copy nodes in the case where underlying data can be modified during export. Got it, thanks.

**Comment by lassoan** [comment]
Let us know if you have addressed all comments and the code is ready for review. Thank you!

**Comment by ebrahimebrahim** [comment]
> Let us know if you have addressed all comments and the code is ready for review. Thank you!

Will do-- it's getting there. Thank you for the review and the many helpful comments!

**Comment by ebrahimebrahim** [comment]
> Error is not displayed. For example, set "Directory" to a non-existing drive or read-only location (x:\tmp or C:\Windows) and click Save. The saving obviously fails, but no error is reported. On models, the error is just silent. For volumes, Slicer crashes.

The issue with volumes should be fixed now, but silent failure when trying to save a _model_ to an unwritable destination seems to be a problem with ordinary saving as well.

**Comment by lassoan** [comment]
Following a similar pattern as you have don in "BUG: Fix silent failure of writing models to file" I've hardened mesh writing in Slicer core (capture all errors and report more details). This will cause merge conflicts with your branch, which the easiest to fix by skipping the daf4f08f9c839c7307e149aa45617a9628df7066 commit from your branch.

**Comment by ebrahimebrahim** [comment]
@lassoan Comments above are addressed-- this is ready for another look!

> which the easiest to fix by skipping the [daf4f08](https://github.com/Slicer/Slicer/commit/daf4f08f9c839c7307e149aa45617a9628df7066) commit from your branch.

I will skip it once I rebase to master to squash the "WIP" commits. Thanks for including me as co-author on 6f3b3c82e6c6b8756815569173409f5398a91200 :grin:

File: Modules/Loadable/Segmentations/qSlicerSegmentationsNodeWriter.cxx
Comment (lassoan) [inline Modules/Loadable/Segmentations/qSlicerSegmentationsNodeWriter.cxx:48]: If you are sure they are not used anywhere then please remove them.

**Review by lassoan — COMMENTED** [review]
This is very good, just added a few nitpicks. Please fix them, then squash the commits, and it should be ready to go.

File: Base/QTGUI/qSlicerExportNodeDialog.cxx
Comment (ebrahimebrahim) [reply Base/QTGUI/qSlicerExportNodeDialog.cxx:61]: Don't I? Leaving them uninitialized would put them in a non-deterministic state, right? This can complicate finding bugs related to them later

File: Base/QTGUI/qSlicerExportNodeDialog.cxx
Comment (ebrahimebrahim) [reply Base/QTGUI/qSlicerExportNodeDialog.cxx:556]: Meaning use `qWarning` rather than `qCritical`?

File: Base/QTGUI/qSlicerExportNodeDialog.cxx
Comment (ebrahimebrahim) [reply Base/QTGUI/qSlicerExportNodeDialog.cxx:299]: Ah, this is referring to item 9 in my original post above:
> Somewhat related to above, and this is a problem with the main save dialog too: It's a little weird that we display differently-cased extensions as separate items in the format list (e.g. ".png" and ".PNG" are separate entries), but we only accept an extension when the user types it in lowercase (e.g. trying to use "MRHead.PNG" as a filename can result in something like "MRHead.PNG.png"). It should really be the other way around: only display lowercase in the format dropdown menu, but accept any case the user chooses to type in. Maybe this should be a separate PR/issue.

This is not really an export dialog problem, so I will remove the TODO. I'll open an issue about it later anyway, so the TODO comment need not be in the code here.

File: Base/QTGUI/qSlicerExportNodeDialog.cxx
Comment (lassoan) [reply Base/QTGUI/qSlicerExportNodeDialog.cxx:61]: Objects are initialized by their default constructor. There is no need for any explicit initialization if you just want to use the default constructor. You must initialize simple types (int, pointer, etc.).

File: Base/QTGUI/qSlicerExportNodeDialog.cxx
Comment (lassoan) [reply Base/QTGUI/qSlicerExportNodeDialog.cxx:556]: Yes. qCritical means that we know that things will not work correctly. qWarning means things may or may not work correctly.

File: Base/QTGUI/qSlicerExportNodeDialog.cxx
Comment (ebrahimebrahim) [reply Base/QTGUI/qSlicerExportNodeDialog.cxx:61]: Ah, you're just talking about the objects. Okay, makes sense.

File: Base/QTGUI/qSlicerExportNodeDialog.cxx
Comment (lassoan) [reply Base/QTGUI/qSlicerExportNodeDialog.cxx:299]: I agree that file extension should be recognized regardless of case. At most places the case insensitive comparison is used, I'm not sure why it is not done here.

Not letting the user choose an arbitrary file extension is a very useful feature. I don't think the user should be allowed to save a nrrd file without extension or .nii file extension.

Anyway, these are all quite harmless, so fixing them would be low priority.

File: Modules/Loadable/Segmentations/qSlicerSegmentationsNodeWriter.cxx
Comment (ebrahimebrahim) [reply Modules/Loadable/Segmentations/qSlicerSegmentationsNodeWriter.cxx:48]: I think the class `qSlicerSegmentationsNodeWriterPrivate` is not getting used anywhere. Should I remove it?

Comment (ebrahimebrahim) [reply Modules/Loadable/Segmentations/qSlicerSegmentationsNodeWriter.cxx]: I removed it

**Comment by ebrahimebrahim** [comment]
@lassoan Additional comments are all addressed now, and commits are squashed.

However I caught an issue: The export dialog was looking for user messages in the storage node of the original node being exported. This is the wrong place to look for errors, since that storage node is not used for writing. Instead, user messages should be collected from the temporary node at the level of `qSlicerCoreIOManager::exportNodes`.

I didn't catch this until I saw the situation that was pointed out in your fix [here](https://github.com/Slicer/Slicer/blob/master/Libs/MRML/Core/vtkMRMLModelStorageNode.cxx#L716). This is a weird situation where the chain of "write" type functions all return a success flag, but there is an indication of failure hiding in the user messages.

Now user messages are properly collected from the temporary node that is used for export. This is done in a separate WIP commit 19adb3c872ba260ea571695bd670d398831ab148 to make it easier for you see the changes. Once approved, I will squash it in with the other ENH commit.

**Comment by ebrahimebrahim** [comment]
> I agree, please move up this check into the storage node so that if any error messages are found then the storage node returns success=0.

Good, I was hoping you'd suggest something like this!
This is now done in 5a2d448ed5e24e9a37f319632b0e8cab075f397a
Note that I changed the documentation for `vtkMRMLStorageNode::WriteData` to indicate that it is not meant to be reimplemented. (Should it then not be virtual?)
Currently, I believe there are no subclasses of `vtkMRMLStorageNode` that try to reimplement `WriteData`.

**Comment by ebrahimebrahim** [comment]
@lassoan Additional comments are all addressed now.

Once approved I will squash 5c1b42af2f522118c642b444d0e66f06ee7844ff into the main ENH commit, and I will leave 5a2d448ed5e24e9a37f319632b0e8cab075f397a as its own "BUG" commit, rather than "WIP." I think it's good to keep 5a2d448ed5e24e9a37f319632b0e8cab075f397a separated.

**Comment by lassoan** [comment]
I think it is all good, please squash the commits.

**Comment by ebrahimebrahim** [comment]
Squashed and rebased to master.

But wait, I'm getting a memory leak when I try to export or save a model as `.ply` to an non-writable path.

```txt
vtkDebugLeaks has detected LEAKS!
Class "vtkHeap" has 1 instance still around.
```

Before merging, let me check that this was not introduced by my changes here

**Comment by lassoan** [comment]
> But wait, I'm getting a memory leak when I try to export or save a model as `.ply` to an non-writable path.
> 
> ```
> vtkDebugLeaks has detected LEAKS!
> Class "vtkHeap" has 1 instance still around.
> ```
> 
> Before merging, let me check that this was not introduced by my changes here

It's OK, it is a bug in VTK and I've already submitted a pull request to fix it (https://gitlab.kitware.com/vtk/vtk/-/merge_requests/8688). The problem will go away in Slicer when the fix is integrated into VTK and Slicer switches to a VTK version that includes it.

**Comment by ebrahimebrahim** [comment]
> It's OK, it is a bug in VTK and I've already submitted a pull request to fix it (https://gitlab.kitware.com/vtk/vtk/-/merge_requests/8688). The problem will go away in Slicer when the fix is integrated into VTK and Slicer switches to a VTK version that includes it.

Ah, okay, awesome. Then this should be good to go once CI is done!

**Comment by lassoan** [comment]
The code is good, now there is just one problem - this commit message: https://github.com/Slicer/Slicer/pull/5876/commits/b0508ce5fd062f9caca7762ed741f035b4e37066

Instead of this long list of detailed and now unusable commit messages (without any idea of what changes each message is related to; and overall too detailed anyway) we just need a concise summary of what was done and why.

At the end you can describe the limitation (no multiple nodes support) and what would need to be done to enable that. It would be nice if you also added a note in the source code (if not there already) where the code should be changed to add support for multiple nodes.

I'm not sure if github can understand `This closes #4989`. Instead, add `fixes $4989` to the end of the commit message.

**Comment by ebrahimebrahim** [comment]
> Instead of this long list of detailed and now unusable commit messages...

Agree, commit message now updated.

> It would be nice if you also added a note in the source code (if not there already) where the code should be changed to add support for multiple nodes.

I think the code changes would be too dispersed for there to be a suitable place to put this note. To be honest, I'm also not crystal-clear on what those changes would be. But I did try to be a little more descriptive in the updated commit message.

**Review by lassoan — APPROVED** [review]

**Comment by lassoan** [comment]
Thank you for all your hard work on this. It is now merged! 🍾 

Please check the dashboard tomorrow to make sure that everything looks as expected. If all looks good then please post a new topic on the [Slicer Forum](https://discourse.slicer.org) that announces the feature to users, providing at least a few-sentence overview and a screenshot. Use `Support` category and add the `feature` keyword (that indicates that it is a new feature announcement).

**Comment by ebrahimebrahim** [comment]
:clinking_glasses: !

---

Looks like a new test is failing [here](https://github.com/Slicer/Slicer/blob/master/Modules/Loadable/Annotations/Testing/Cxx/vtkMRMLAnnotationStorageNodeTest1.cxx#L47)... I'm guessing there was something wrong with that test to begin with, and it's now able to correctly report failure due to the error-handling changes introduced in `vtkMRMLStorageNode` in c866c3959bd5f02746d0c6cc92ec902e3e777a27

**Pull request merged**
